src/linalg/Lanczos_Exp.cpp

```diff
@@ -0,0 +1,227 @@
+#include "linalg.hpp"
+#include "Generator.hpp"
+#include "random.hpp"
+#include "Tensor.hpp"
+#include "LinOp.hpp"
+
+#include <cfloat>
+#include <vector>
+#include <cmath>
+#include "UniTensor.hpp"
+#include "utils/vec_print.hpp"
+#include <iomanip>
+
+#ifdef BACKEND_TORCH
+#else
+
+namespace cytnx {
+  namespace linalg {
+    typedef Accessor ac;
+    using namespace std;
+
+    // <A|B>
+    static Scalar Dot(const UniTensor &A, const UniTensor &B) {
+      return Contract(A.Dagger(), B).item();
+    }
+
+    // BiCGSTAB method to solve the linear equation
+    // ref: https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method
+    UniTensor Invert_BiCGSTAB(LinOp *Hop, const UniTensor &b, const UniTensor &Tin, const int &k,
+                              const double &CvgCrit = 1.0e-12,
+                              const unsigned int &Maxiter = 10000) {
+      auto I_plus_A_Op = [&](UniTensor A) {
+        return ((Hop->matvec(A)) / k + A).relabels_(b.labels());
+      };
+      auto r = (b - I_plus_A_Op(Tin)).relabels_(b.labels());
+      auto r0 = r;
+      auto x = Tin;
+      auto p = Dot(r0, r);
+      auto pv = r;
+      auto p_old = p;
+      auto pv_old = pv;
+      auto x_old = x;
+      auto r_old = r;
```

Comment:
Why is this so scattered? maybe we should considered grouping parameters into a struct